fix(ivy): several small fixes to run cli-hello-world-ivy integration test suite

JoostK opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

The cli-hello-world-ivy test would only build, not actually run. With some Ivy fixes the tests can now actually be enabled:

**Fixes:**
- Pipes would not be available to JIT compiled modules.
- Several `__POST_R3__` symbols were DCE'd. The integration test now finally catches those.
- The `setClassMetadata` would fail to be generated from ngcc for ES5 bundles.

This builds on top of #27159 as the ESM5 bundle is used for tests.

## What is the new behavior?

Ivy Hello World integration test actually runs :tada:

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JoostK committed:
fix(ivy): prevent DCE of __POST_R3__ symbols

While creating FESM files, rollup usually drops all unused symbols.
All *__POST_R3__ are unused unless ngcc rewires stuff. To prevent this DCE
we reexport them as private symbols. If ngcc is not used, these symbols will
be dropped when we optimize an application bundle.

JoostK committed:
fix(ivy): make exported pipes available in module scope

This is tested in the Ivy CLI integration test.

JoostK committed:
test(ivy): synchronize cli-hello-world-ivy with cli-hello-world integration suite

The ci-production e2e test is diabled because Angular fails to be found

JoostK committed:
test(ivy): enable running cli-hello-world-ivy integration test suite

JoostK commented on packages/core/src/render3/jit/module.ts:
```diff
@@ -51,6 +51,10 @@ export function compileNgModuleDefs(moduleType: Type<any>, ngModule: NgModule):
               emitInline: true,
             });
+
+        if (applyScope) {
```
This is a bit of a hack, but `TestBed` applies its own module scopes which this would otherwise interfere with. I have some work in #27399 to hopefully avoid this in the future, but that needs some more work.

JoostK commented on packages/core/src/render3/jit/module.ts:
```diff
@@ -198,7 +202,7 @@ export function transitiveScopesFor<T>(moduleType: Type<T>): NgModuleTransitiveS
         scopes.exported.pipes.add(entry);
       });
-    } else if (getNgModuleDef(exportedTyped)) {
+    } else if (getPipeDef(exportedTyped)) {
```
This bug was not found with `TestBed` tests as it has its own idea of module scopes, so this code was not relevant there. With the work in #27399 I deleted the custom module scope application (for now, we might still need something similar as my current approach seems too limited) and that uncovered this issue.

JoostK commented:
I'll have to investigate why Travis fails with `Error: Importing BrowserDynamicTestingModule which does not have an @ngModule`. Somehow the `setClassMetaData` call is not present in the FESM5 bundles. I have seen this working locally but now also have the same error, not sure what has changed 🤷‍♂️

gkalpak reviewed (commented):
I have no idea about the `setScopeOnDeclaredComponents()` change.
The rest looks great :smiley:

pkozlowski-opensource commented:
@JoostK @gkalpak I think that `TestBed` is definitively doing sth fishy with scopes, I remember discussing it briefly with @alxhub in the past.

JoostK commented:
> @JoostK @gkalpak I think that `TestBed` is definitively doing sth fishy with scopes, I remember discussing it briefly with @alxhub in the past.

Yes, it definitely is. Basically `TestBed` has its own module scoping as it walks the module graph to recompile everything anyway based on the registered overrides. That's what #27399 is supposed to avoid, although the module scope modifications prove to be problematic.

JoostK committed:
fix(ivy): ngcc generates setClassMetadata calls for ES5 bundles

ngcc would feed ngtsc with the function declaration inside of an IIFE as
that is considered the class symbol's declaration node, according to
TypeScript's `ts.Symbol.valueDeclaration`. ngtsc however only considered
variable decls and actual class decls as potential class declarations,
so given the function declaration node it would fail to generate the
`setClassMetadata` call.

ngtsc no longer makes its own assumptions about what classes look like,
but always asks the reflection host to yield this kind of information.

JoostK committed:
perf(ivy): let ngcc first check marker file before assembling bundle

With the bundle info being assembled into a single object before the
transform is started, we now greedily create a TypeScript program up-front.
If a marker file exists that indicates that the bundle could be skipped
the program creation has already taken place which takes a significant
amount of time. This commit moves the marker check to occur before the
bundle is assembled.

JoostK commented:
@gkalpak All green now 🍏 💚  I've had to disable the ci-production e2e configuration as that would timeout with an error that Angular cannot be found. Don't know what's going on exactly but I'm accepting for now that that test just doesn't run yet.

gkalpak commented on packages/compiler-cli/src/ngcc/src/main.ts:
```diff
@@ -59,6 +60,11 @@ export function mainNgcc(args: string[]): number {
+        if (checkMarkerFile(entryPoint, format)) {
+          console.warn(`Skipping ${entryPoint.name} : ${format} (already built).`);
+          return;
+        }
```
👌

gkalpak commented on integration/cli-hello-world-ivy/package.json:
```diff
@@ -3,13 +3,13 @@
+    "postinstall": "webdriver-manager update --gecko=false --standalone=false $CHROMEDRIVER_VERSION_ARG && yarn ivy-ngcc",
     "start": "ng serve",
-    "test": "ng build --progress false"
+    "test": "ng test --progress=false --watch=false && yarn e2e --configuration=ci"
```
Can you leave a TODO comment here mentioning that `ci-production` is failing? E.g.:

```json
"//test1": "TODO: Re-enable `ci-production`. Currently, it fails like this and that.", 
"//test2": "ng test --progress=false --watch=false && yarn e2e --configuration=ci && yarn e2e --configuration=ci-production",
```

JoostK replied:
@gkalpak Good idea, done!

gkalpak replied:
Thx! Can you also add that this is tracked in Jira issue is FW-813?

gkalpak commented:
Marking this as blocked on #27159. (I'd prefer to merge that first in order to preserve the PR comment history.)

alxhub commented:
Nice work!

alxhub reviewed (approved):
Waiting to rebase once #27159 lands

JoostK commented:
Rebased now that #27159 has been merged. Tests are green, presubmit time!

gkalpak commented:
Note to caretaker: Assistance needed with g3sync.
